proposals/p5337.md

```diff
@@ -0,0 +1,967 @@
+
+```carbon
+interface A {
+  fn F();
+}
+
+interface B {
+  extend require A;
+  fn G();
+}
+```
+
+is equivalent to:
+
+```carbon
+interface A {
+  fn F();
+}
+
+interface B {
+  require Self impls A;
+  alias F = A.F;
+  fn G();
+}
+```
+
+> FIXME _**QUESTION**_: Is it `extend require I` or `extend require impls I`?
+
+> FIXME _**QUESTION**_: Should we allow `require impls I` as a shortcut for
+> `require Self impls I` at the same time?
+
+Next, we define the meaning of `extend impl as I;` in an `interface J` scope to
+do two things:
+
+-   Copy the members of `I` to form new members of `J`.
+-   Define a blanket `impl` that anything that implements `J` also implements
+    `I`, by forwarding to the corresponding members of `J`.
+
+In this example:
+
+```carbon
+interface I {
+  let T:! type;
+  fn F();
+  fn G();
+}
+
+interface J {
+  extend impl as I;
+  fn H();
+}
+```
+
+The definition of `interface J` is equivalent to:
+
+```carbon
+interface J {
+  let T:! type;
+  fn F();
+  fn G();
+  fn H();
+}
+impl forall [U:! J] U as I {
+  let T:! type = U.(J.T);
+  fn F() = U.(J.F);
+  fn G() = U.(J.G);
+}
+```
+
+The keyword `final` can be added between `extend` and `impl` to make the
+generated `impl` definition `final`. For example, the
+`final impl forall [T:! type, U:! ImplicitAs(T)] U as As(T)` from
+[the "Problem" section](#problem) would be generated from this:
+
+```carbon
+interface ImplicitAs(T:! type) {
+  extend final impl as As(T);
+}
+```
+
+To allow an `impl` to overlap a final `impl`, we say the final `impl` is
+prioritized over a non-final `impl` anytime they overlap. This addresses a
+[problem](#problem) caused by the fact that we don't support the negative
+constraints that would allow a developer to avoid the overlap -- and the only
+sound choice is to prefer the final `impl`. However, if an `impl` can never be
+used because it is completely subsumed by a final `impl`, that is an error.
+
+To allow multiple interfaces extending the same interface, we also want to
+support a way to declare overlapping final impls. In this case, list the impls
+in a `final match_first` block instead of putting `final` on the `impl`
+declaration itself. To put the impl generated by an `extend impl as Foo` in an
+interface `Bar` in a `match_first` block (or a `final match_first` block), we
+give it the name `impl Bar.(as Foo)`.
+
+This example inspired by
+[C++ iterator categories](https://en.cppreference.com/w/cpp/iterator) shows
+using these pieces together:
+
+```carbon
+interface Iterator {
+  fn Increment[addr self: Self*]();
+}
+
+interface InputIterator {
+  let Element:! type;
+  extend impl as Iterator;
+  fn Get[addr self: Self*]() -> Element;
+}
+
+interface OutputIterator {
+  let Element:! type;
+  extend impl as Iterator;
+  fn Set[addr self: Self*](x: Element);
+}
+
+// Makes both impls final and prioritizes them.
+final match_first {
+  impl InputIterator.(as Iterator);
+  impl OutputIterator.(as Iterator);
+}
```

Comment:
Potential problem with this approach: if I have `T:! OutputIterator where .Element = i32`, I know that `T impls Iterator` but I do not know that `T.(Iterator.Element) == i32`. Specifically, if `T` also implements `InputIterator` with a different `.Element`, that `.Element` wins.

Maybe it'd be better to handle this overlap by making all the `extend impl`s final and adding this to the library:
```carbon
interface InputOutputIterator {
  extend final impl as InputIterator;
  extend final impl as OutputIterator;
}
```
... so that a type that is both an input iterator and an output iterator can still have only one element type.

Reply:
Also, we can't make two interfaces `extend final impl as` a single common interface.